models/mobilenet/mobilenet_v1_impl.hpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * @file mobilenet_v1_impl.hpp
+ * @author Aakash Kaushik
+ *
+ * Implementation of MobileNetV1 using mlpack.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MODELS_MODELS_MOBILENET_MOBILENET_V1_IMPL_HPP
+#define MODELS_MODELS_MOBILENET_MOBILENET_V1_IMPL_HPP
+
+#include "mobilenet_v1.hpp"
+
+namespace mlpack {
+namespace models {
+
+template<typename OutputLayerType, typename InitializationRuleType>
+MobileNetV1<OutputLayerType, InitializationRuleType>::MobileNetV1() :
+    inputChannel(0),
+    inputWidth(0),
+    inputHeight(0),
+    numClasses(0),
+    alpha(0),
+    depthMultiplier(0)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
+MobileNetV1<OutputLayerType, InitializationRuleType>::MobileNetV1(
+    const size_t inputChannel,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const size_t alpha,
+    const size_t depthMultiplier,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    MobileNetV1<OutputLayerType, InitializationRuleType>(
+        std::tuple<size_t, size_t, size_t>(
+        inputChannel,
+        inputWidth,
+        inputHeight),
+        alpha,
+        depthMultiplier,
+        includeTop,
+        preTrained,
+        numClasses)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
+MobileNetV1<OutputLayerType, InitializationRuleType>::MobileNetV1(
+    std::tuple<size_t, size_t, size_t> inputShape,
+    const size_t alpha,
+    const size_t depthMultiplier,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    inputChannel(std::get<0>(inputShape)),
+    inputWidth(std::get<1>(inputShape)),
+    inputHeight(std::get<2>(inputShape)),
+    alpha(alpha),
+    depthMultiplier(depthMultiplier),
+    numClasses(numClasses)
+{
+  if (inputWidth < 32 || inputHeight < 32)
+  {
+    mlpack::Log::Fatal << "input width and input height cannot be smaller than"
+        " 32.\nGiven input width and height: (" << inputWidth << ", "
+        << inputHeight << ")" << std::endl;
+  }
+  outSize = size_t(32 * alpha);
+  mobileNet.Add(new ann::Convolution<>(inputChannel, outSize, 3, 3, 2, 2, 1, 1,
```

Comment:
https://github.com/mlpack/mlpack/blob/42071e92be63221da73a620de5beac0a5c537a9c/src/mlpack/methods/ann/layer/convolution_impl.hpp#L71,
Here, total padding would be of 2 but in the PyTorch implementation, https://github.com/ZFTurbo/MobileNet-v1-Pytorch/blob/e08a77a1b93f7fcb68960766cf5958cee3c1a134/mobilenet_v1.py#L19, could you see if dims match?

Reply:
Hey, so as you can see the output dimensions do match. 
Mlpack
```bash
[INFO ] Convolution: (3, 224, 224) ---> (32, 112, 112)
[INFO ] BatchNorm: (32) ---> (32)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Separable convolution: (32, 112, 112) ---> (32, 112, 112)
[INFO ] BatchNorm: (32) ---> (32)
[INFO ] RelU6
[INFO ] Convolution: (32, 112, 112) ---> (64, 112, 112)
[INFO ] BatchNorm: (64) ---> (64)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Padding: (64, 112, 112 ---> (64, 113, 113)
[INFO ] Separable convolution: (64, 113, 113) ---> (64, 56, 56)
[INFO ] BatchNorm: (64) ---> (64)
[INFO ] RelU6
[INFO ] Convolution: (64, 56, 56) ---> (128, 56, 56)
[INFO ] BatchNorm: (128) ---> (128)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Separable convolution: (128, 56, 56) ---> (128, 56, 56)
[INFO ] BatchNorm: (128) ---> (128)
[INFO ] RelU6
[INFO ] Convolution: (128, 56, 56) ---> (128, 56, 56)
[INFO ] BatchNorm: (128) ---> (128)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Padding: (128, 56, 56 ---> (128, 57, 57)
[INFO ] Separable convolution: (128, 57, 57) ---> (128, 28, 28)
[INFO ] BatchNorm: (128) ---> (128)
[INFO ] RelU6
[INFO ] Convolution: (128, 28, 28) ---> (256, 28, 28)
[INFO ] BatchNorm: (256) ---> (256)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Separable convolution: (256, 28, 28) ---> (256, 28, 28)
[INFO ] BatchNorm: (256) ---> (256)
[INFO ] RelU6
[INFO ] Convolution: (256, 28, 28) ---> (256, 28, 28)
[INFO ] BatchNorm: (256) ---> (256)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Padding: (256, 28, 28 ---> (256, 29, 29)
[INFO ] Separable convolution: (256, 29, 29) ---> (256, 14, 14)
[INFO ] BatchNorm: (256) ---> (256)
[INFO ] RelU6
[INFO ] Convolution: (256, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Separable convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Separable convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Separable convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Separable convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Separable convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Convolution: (512, 14, 14) ---> (512, 14, 14)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Padding: (512, 14, 14 ---> (512, 15, 15)
[INFO ] Separable convolution: (512, 15, 15) ---> (512, 7, 7)
[INFO ] BatchNorm: (512) ---> (512)
[INFO ] RelU6
[INFO ] Convolution: (512, 7, 7) ---> (1024, 7, 7)
[INFO ] BatchNorm: (1024) ---> (1024)
[INFO ] RelU6
[INFO ] Block starts: 
[INFO ] Separable convolution: (1024, 7, 7) ---> (1024, 7, 7)
[INFO ] BatchNorm: (1024) ---> (1024)
[INFO ] RelU6
[INFO ] Convolution: (1024, 7, 7) ---> (1024, 7, 7)
[INFO ] BatchNorm: (1024) ---> (1024)
[INFO ] RelU6
[INFO ] Mean pooling: (1024, 7, 7) ---> (1024, 1, 1)
[INFO ] Dropout
[INFO ] Convolution: (1024, 1, 1) ---> (1000 , 1, 1)
[INFO ] Softmax
```
PyTorch
```bash
---------------------------------------------------------------
        Layer (type)               Output Shape         Param #
================================================================
     ConstantPad2d-1          [-1, 3, 225, 225]               0
            Conv2d-2         [-1, 32, 112, 112]             864
       BatchNorm2d-3         [-1, 32, 112, 112]              64
             ReLU6-4         [-1, 32, 112, 112]               0
            Conv2d-5         [-1, 32, 112, 112]             288
       BatchNorm2d-6         [-1, 32, 112, 112]              64
             ReLU6-7         [-1, 32, 112, 112]               0
            Conv2d-8         [-1, 64, 112, 112]           2,048
       BatchNorm2d-9         [-1, 64, 112, 112]             128
            ReLU6-10         [-1, 64, 112, 112]               0
    ConstantPad2d-11         [-1, 64, 113, 113]               0
           Conv2d-12           [-1, 64, 56, 56]             576
      BatchNorm2d-13           [-1, 64, 56, 56]             128
            ReLU6-14           [-1, 64, 56, 56]               0
           Conv2d-15          [-1, 128, 56, 56]           8,192
      BatchNorm2d-16          [-1, 128, 56, 56]             256
            ReLU6-17          [-1, 128, 56, 56]               0
           Conv2d-18          [-1, 128, 56, 56]           1,152
      BatchNorm2d-19          [-1, 128, 56, 56]             256
            ReLU6-20          [-1, 128, 56, 56]               0
           Conv2d-21          [-1, 128, 56, 56]          16,384
      BatchNorm2d-22          [-1, 128, 56, 56]             256
            ReLU6-23          [-1, 128, 56, 56]               0
    ConstantPad2d-24          [-1, 128, 57, 57]               0
           Conv2d-25          [-1, 128, 28, 28]           1,152
      BatchNorm2d-26          [-1, 128, 28, 28]             256
            ReLU6-27          [-1, 128, 28, 28]               0
           Conv2d-28          [-1, 256, 28, 28]          32,768
      BatchNorm2d-29          [-1, 256, 28, 28]             512
            ReLU6-30          [-1, 256, 28, 28]               0
           Conv2d-31          [-1, 256, 28, 28]           2,304
      BatchNorm2d-32          [-1, 256, 28, 28]             512
            ReLU6-33          [-1, 256, 28, 28]               0
           Conv2d-34          [-1, 256, 28, 28]          65,536
      BatchNorm2d-35          [-1, 256, 28, 28]             512
            ReLU6-36          [-1, 256, 28, 28]               0
    ConstantPad2d-37          [-1, 256, 29, 29]               0
           Conv2d-38          [-1, 256, 14, 14]           2,304
      BatchNorm2d-39          [-1, 256, 14, 14]             512
            ReLU6-40          [-1, 256, 14, 14]               0
           Conv2d-41          [-1, 512, 14, 14]         131,072
      BatchNorm2d-42          [-1, 512, 14, 14]           1,024
            ReLU6-43          [-1, 512, 14, 14]               0
           Conv2d-44          [-1, 512, 14, 14]           4,608
      BatchNorm2d-45          [-1, 512, 14, 14]           1,024
            ReLU6-46          [-1, 512, 14, 14]               0
           Conv2d-47          [-1, 512, 14, 14]         262,144
      BatchNorm2d-48          [-1, 512, 14, 14]           1,024
            ReLU6-49          [-1, 512, 14, 14]               0
           Conv2d-50          [-1, 512, 14, 14]           4,608
      BatchNorm2d-51          [-1, 512, 14, 14]           1,024
            ReLU6-52          [-1, 512, 14, 14]               0
           Conv2d-53          [-1, 512, 14, 14]         262,144
      BatchNorm2d-54          [-1, 512, 14, 14]           1,024
            ReLU6-55          [-1, 512, 14, 14]               0
           Conv2d-56          [-1, 512, 14, 14]           4,608
      BatchNorm2d-57          [-1, 512, 14, 14]           1,024
            ReLU6-58          [-1, 512, 14, 14]               0
           Conv2d-59          [-1, 512, 14, 14]         262,144
      BatchNorm2d-60          [-1, 512, 14, 14]           1,024
            ReLU6-61          [-1, 512, 14, 14]               0
           Conv2d-62          [-1, 512, 14, 14]           4,608
      BatchNorm2d-63          [-1, 512, 14, 14]           1,024
            ReLU6-64          [-1, 512, 14, 14]               0
           Conv2d-65          [-1, 512, 14, 14]         262,144
      BatchNorm2d-66          [-1, 512, 14, 14]           1,024
            ReLU6-67          [-1, 512, 14, 14]               0
           Conv2d-68          [-1, 512, 14, 14]           4,608
      BatchNorm2d-69          [-1, 512, 14, 14]           1,024
            ReLU6-70          [-1, 512, 14, 14]               0
           Conv2d-71          [-1, 512, 14, 14]         262,144
      BatchNorm2d-72          [-1, 512, 14, 14]           1,024
            ReLU6-73          [-1, 512, 14, 14]               0
    ConstantPad2d-74          [-1, 512, 15, 15]               0
           Conv2d-75            [-1, 512, 7, 7]           4,608
      BatchNorm2d-76            [-1, 512, 7, 7]           1,024
            ReLU6-77            [-1, 512, 7, 7]               0
           Conv2d-78           [-1, 1024, 7, 7]         524,288
      BatchNorm2d-79           [-1, 1024, 7, 7]           2,048
            ReLU6-80           [-1, 1024, 7, 7]               0
           Conv2d-81           [-1, 1024, 7, 7]           9,216
      BatchNorm2d-82           [-1, 1024, 7, 7]           2,048
            ReLU6-83           [-1, 1024, 7, 7]               0
           Conv2d-84           [-1, 1024, 7, 7]       1,048,576
      BatchNorm2d-85           [-1, 1024, 7, 7]           2,048
            ReLU6-86           [-1, 1024, 7, 7]               0
        AvgPool2d-87           [-1, 1024, 1, 1]               0
           Conv2d-88           [-1, 1000, 1, 1]       1,025,000
================================================================
```